src/coreclr/jit/emit.cpp

```diff
@@ -1021,27 +1021,28 @@ insGroup* emitter::emitSavIG(bool emitAdd)
     }
 
     // Fix the last instruction field, if set. Note that even if there are instructions in the IG,
-    // emitLastIns might not be set if an optimization has just deleted it, and the new instruction
-    // being adding causes a new EXTEND IG to be created. Also, emitLastIns might not be in this IG
+    // emitGetLastIns might not be set if an optimization has just deleted it, and the new instruction
+    // being adding causes a new EXTEND IG to be created. Also, emitGetLastIns might not be in this IG
     // at all if this IG is empty.
 
-    assert((emitLastIns == nullptr) == (emitLastInsIG == nullptr));
-    if ((emitLastIns != nullptr) && (sz != 0))
+    assert(emitHasLastIns() == (emitLastInsIG != nullptr));
+    if (emitHasLastIns() && (sz != 0))
     {
-        // If we get here, emitLastIns must be in the current IG we are saving.
+        // If we get here, emitGetLastIns must be in the current IG we are saving.
         assert(emitLastInsIG == emitCurIG);
-        assert(emitCurIGfreeBase <= (BYTE*)emitLastIns);
-        assert((BYTE*)emitLastIns < emitCurIGfreeBase + sz);
+        assert(emitCurIGfreeBase <= (BYTE*)emitGetLastIns());
+        assert((BYTE*)emitGetLastIns() < emitCurIGfreeBase + sz);
 
 #if defined(TARGET_XARCH)
-        if (emitLastIns->idIns() == INS_jmp)
+        if (emitGetLastIns()->idIns() == INS_jmp)
         {
             ig->igFlags |= IGF_HAS_REMOVABLE_JMP;
         }
 #endif
 
-        emitLastIns   = (instrDesc*)((BYTE*)id + ((BYTE*)emitLastIns - (BYTE*)emitCurIGfreeBase));
-        emitLastInsIG = ig;
+        emitLastInstrs[0] = (instrDesc*)((BYTE*)id + ((BYTE*)emitGetLastIns() - (BYTE*)emitCurIGfreeBase));
+        emitCurLastInsCnt = 1;
```

Comment:
I think there's a fundamental problem with capping the last instruction buffer when we hit an IG boundary: we will get Checked/Release asm diffs.

The problem is that the IG buffer of instrDescs is filled with different numbers of instructions in Checked and Release and so different optimizations will occur when that boundary is crossed.

I think you need to update all the saved instrs if `emitAdd` is true. If it is false, you can cap it to one, as done here.

Reply:
We have to work across IGF_EXTEND IG boundaries for correctness. TP needs to be considered in that context.

Reply:
Ok, I will add back the update loop.